docs/rfc/rfc-004-accounts.md

```diff
@@ -0,0 +1,323 @@
+# RFC 004: Accounts
+
+## Changelog
+
+- 17/03/2023: DRAFT
+- 09/05/2023: DRAFT 2
+
+## Context
+
+The current implementation of accounts in the Cosmos SDK is limiting in terms of functionality, extensibility, and overall
+architecture. This RFC aims to address the following issues with the current account system:
+
+### 1. Accounts Representation and Authentication Mechanism
+
+The current SDK accounts are represented as `google.Protobuf.Any`, which are then encapsulated into the account interface.
+This account interface essentially represents the authentication mechanism, as it implements methods such as `GetNumber`
+and `GetSequence` that serve as abstractions over the authentication system. However, this approach restricts the scope and
+functionality of accounts within the SDK.
+
+### 2. Limited Account Interface
+
+The account interface in its current form is not versatile enough to accommodate more advanced account functionalities,
+such as implementing vesting capabilities or more complex authentication and authorization systems.
+
+### 3. Multiple Implementations of the Account Interface
+
+There are several implementations of the account interface, like `ModuleAccount`, but the existing abstraction does not
+allow for meaningful differentiation between them. This hinders the ability to create specialized accounts that cater to
+specific use cases.
+
+### 4. Primitive Authorization System
+
+The authorization system in the `x/auth` module is basic and defines authorizations solely for the functionalities of the
+`x/bank` module. Consequently, although the state transition authorization system is defined in `x/auth`, it only covers the
+use cases of `x/bank`, limiting the system's overall scope and adaptability.
+
+### 5. Cyclic Dependencies and Abstraction Leaks
+
+The current account system leads to cyclic dependencies and abstraction leaks throughout the Cosmos SDK. For instance,
+the `Vesting` functionality belongs to the `x/auth` module, which depends on the `x/bank` module. However,
+the `x/bank` module depends on the `x/auth` module again to identify the account type (either `Vesting` or `Base`) during
+a coin transfer. This dependency structure creates architectural issues and complicates the overall design of the SDK.
+
+## Proposal
+
+This proposal aims to transform the way accounts are managed within the Cosmos SDK by introducing significant changes to
+their structure and functionality.
+
+### Rethinking Account Representation and Business Logic
+
+Instead of representing accounts as simple `google.Protobuf.Any` structures stored in state with no business logi
+attached, this proposal suggests a more sophisticated account representation that is closer to module entities.
+In fact, accounts should be able to receive messages and process them in the same way modules do, and be capable of storing
+state in a isolated (prefixed) portion of state belonging only to them, in the same way as modules do.
```

Comment:
I hadn't noticed this bit before, but this is great! It allows accounts to modify that part of the state regardless of whether the msg succeeds or fails authorization. It's still a bit unclear where this would be enforced, but that's likely an implementation detail